### PR TITLE
Add support for generic URI in Storage object 

### LINKED ISSUE #826: Support for general web URI address for hosted KFServing model location
/kind feature

**Describe the solution you'd like**

Currently there are 4 types of storage URIs that are accepted in KFServing yaml files:

https://github.com/kubeflow/kfserving/blob/e25758f6da020e3e8ad63f0880ab95de6af8395b/python/kfserving/kfserving/storage.py#L26-L30

(1) GCS
(2) AWS S3
(3) Azure (blob over https://)
(4) local (file://)

I think it would be useful to be able reference a model file hosted anywhere (via general web URI, aka "http://", such as a file on github:

https://github.com/jpatanooga/kubeflow_ops_book_dev/blob/master/kfserving/models/sklearn/iris_sklearn_svm_model.joblib

(in this example the raw file would be downloaded from the url: https://github.com/jpatanooga/kubeflow_ops_book_dev/blob/master/kfserving/models/sklearn/iris_sklearn_svm_model.joblib?raw=true )

This would allow for examples to be executed more efficiently; the KFServing project could even provide a repo of the pre-built models (perhaps), so that example model deployments could be easier to run for new users.

### PR DESCRIPTION
**Pull request by tduffy000**
**What this PR does / why we need it**:
This adds support for a generic uri over HTTP or HTTPS in the Storage [object](https://github.com/kubeflow/kfserving/blob/91165ea3b94f77432029ecc54bec7f4f888f842b/python/kfserving/kfserving/storage.py#L32) responsible for pulling remote model objects, as mentioned in this [issue](https://github.com/kubeflow/kfserving/issues/826). This would allow for the inclusion of serialized models available on the web (e.g. on GitHub, ModelZoo, etc.) in the KFServing framework.

Re-opens https://github.com/kubeflow/kfserving/pull/972.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #826

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by tduffy000** [comment]
I will add a sample [here](https://github.com/kubeflow/kfserving/tree/master/docs/samples#deploy-inferenceservice-with-cloudpvc-storage), as requested in the previous [review](https://github.com/kubeflow/kfserving/pull/972#issuecomment-664063720).

File: python/kfserving/kfserving/storage.py [hunk 1]
```diff
@@ -206,6 +211,18 @@ def _download_local(uri, out_dir=None):
             os.symlink(src, dest_path)
         return out_dir
 
+    @staticmethod
+    def _download_from_uri(uri, out_dir=None):
+        url = urlparse(uri)
+        fname = os.path.basename(url.path)
```
Comment (tduffy000) [inline python/kfserving/kfserving/storage.py:217]: should we validate whether this is `''` and fail loudly? E.g. if the url doesn't contain a valid file name?

Comment (yuzisun) [reply python/kfserving/kfserving/storage.py:217]: sgtm, I guess something like `http://foo.bar/test/` will also fail right?

Comment (tduffy000) [reply python/kfserving/kfserving/storage.py:217]: yes, it should, I'll add that along with an accompanying unit test.

Comment (tduffy000) [reply python/kfserving/kfserving/storage.py:217]: addressed in https://github.com/kubeflow/kfserving/pull/979/commits/00de564cb6a646146896946261a41d9699325a5a

**Comment by tduffy000** [comment]
I'm running into an error when attempting to `apply` any of the manifests in the `samples` directory:
```bash
$ kubectl apply -f docs/samples/
error: error when retrieving current configuration of:
Resource: "serving.kubeflow.org/v1alpha2, Resource=inferenceservices", GroupVersionKind: "serving.kubeflow.org/v1alpha2, Kind=InferenceService"
Name: "sklearn-iris", Namespace: "default"

```
I tried deploying them into a different namespace and got the same error. Similarly, both deploying via `make deploy` and `make deploy-dev` return the same issue. I'm on k8s v1.16.0 and KNative version v0.15.0. I can see the kfserving controller running:
```bash
$ kubectl get pods -n kfserving-system
NAME                             READY   STATUS    RESTARTS   AGE
kfserving-controller-manager-0   2/2     Running   0          ----
```

Attempting to `curl` it (after messing with the RBAC) also returns the same message as above.
```bash
$ curl -k -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.16.0 (linux/amd64) kubernetes/2bd9643" 'https://localhost:8443/apis/serving.kubeflow.org/v1alpha2/namespaces/kfserving-test/inferenceservices/sklearn-iris'
```

**Comment by tduffy000** [comment]
@yuzisun I've hit a bit of a fork in the road here and wondering what you (or others) might think about the best way forward is here. The current implementation will only work for frameworks which require a single model artifact, e.g. sklearn (which I successfully tested). However, other supported frameworks require access to multiple files (like Tensorflow) in order to pull down the serialized model and turn it into a servable.

Which means, pulling down a single serialized object isn't going to cut it. All the other storage representations currently implemented provide some kind of directory listing which makes traverse down the directory structure straightforward, however, this is not really how the web works (most servers aren't going to tell us what the paths are). 

I see two potential solutions for handling all the frameworks `kfserving` supports while still being able to specify any `http(s)` path:
1. Introduce a new dependency like `BeautifulSoup` or `scrapy` and essentially take the base path,
```yaml
spec:
  default:
    predictor:
      tensorflow:
        storageUri: https://someurl/model/
```
and attempt to crawl from the top and downloading all the artifacts under that path. This would have the benefit of maintaining the current structure of the `InferenceService` CRD, but might lead to some undefined behavior that the user _probably_ won't have visibility into unless they are manually checking the objects downloaded by the `storage-initializer` container. 

2. Add a new (optional) field to the CRD like this:
```spec:
  default:
    predictor:
      tensorflow:
        storageUri: https://someurl/model/
          paths:
            - 0001/saved_model.pb
            - 0001/variables/variables.data-00000-of-00001
            - 0001/variables/variables.index
```
while this doesn't explicitly _break_ the current CRD contract of just providing `storageUri` it does introduce a new field (and that would only be required for model objects available from this storage type (that said I guess we could add support for the others). But has the benefit of being obvious to the user what _should_ be downloaded to fully specify the `kfserving-container`. 

That said, if you (or others) have ideas about how to approach this (the two above or others), I'd be interested in hearing them.

**Comment by ellistarn** [comment]
You're completely right -- this is a sticky issue with a conceptual mismatch.

I'm not a fan of the paths approach. Your YAML also isn't valid. Paths would either need to be a sibling to storageUri, or storageUri would need to become an object. Both approaches are not ideal.

I could see using some sort of directory lister, like an index.yaml, scrappy, or other way of listing files.

**Comment by tduffy000** [comment]
I would rather not touch the current CRD structure, so makes sense that the `paths` approach is not desirable.

I have a concern about how general of a solution we can provide here without throwing another config at the problem. As an example, even building a solution that would work on a Github directory, like 
```yaml
...
    storageUri: https://github.com/<owner>/<repo>/<path>/<to>/<serialized>/<model>/
```
is tricky because the raw binaries Github  are not hosted under _that_ url, but under `https://raw.githubusercontent.com/<owner>/<repo>/<path>/<to>/<serialized>/<model>/<object>`. We can eventually get there by crawling the links, but cannot verify that the files are _directly_ under the path the user specified in the CRD. Trying to find a generic enough solution for this, given the possible redirections a site may take us through seems difficult, and has what I'm guessing is an incredible number of edge cases. 

So, I would _think_ that the introduction of declarative directory listing from the user would allow us to maintain the current proposal of accepting _any_ http(s) path. Otherwise, it might be more appropriate to make this more specific to different potential hosts, e.g. `github://<some>/<path>/...`, `gitlab://<some>/<path>`, with the unfortunate side effect of probably bloating the `Storage` class with all the possible options.

**Comment by ellistarn** [comment]
We will need to be a little strategic if we go this route, as the storage initializer component will soon be deprecated in favor of the new puller agent that enables support for multi-model serving. I'm a little nervous about where this road leads. I see storing models in github as your cloud storage provider as something of an anti-pattern.

Is it possible to mirror the github directory into cloud storage somewhere?

**Comment by tduffy000** [comment]
Perhaps if this is being deprecated soon we should close the issue (and this PR). I'm in agreement that using GitHub (or another publicly accessible location) is probably not a valid use case for production-level artifact storage. I didn't open https://github.com/kubeflow/kfserving/issues/826 but my _guess_ is that this would probably only be useful for learning purposes (which certainly shouldn't discounted) or an easy proof-of-concept. 

As to the second question, one would think a simple `sync` command on AWS / GCP / Azure / take your pick in a standard CI/CD pipeline would allow mirroring that structure (even if you are committing artifacts directly to git). I don't have a use case for this (simply picked it up because of the `good-first-issue` label), but someone else might :shrug: .

**Comment by yuzisun** [comment]
@tduffy000 @ellistarn The original issue is for learning purpose, not everyone has a proper artifact storage. I think the PR still has value and I am not sure if we should deprecate storage initializer yet for single model serving case. Perhaps we can limit the scope to only allow single file or tarball if multiple model files are needed, it can also be used for e2e testing purpose.

**Comment by tduffy000** [comment]
I'm okay with that, I think it will be a nice way for folks without dedicated storage to get a feel for the framework. I will add support for `.zip` and `.tar` files (along with a nice warning about only pulling those in from *trusted* sources) along with samples for each framework (given that some will require artifacts and others won't).

**Comment by ellistarn** [comment]
I like this tar support! Nice!

**Comment by tduffy000** [comment]
@yuzisun @ellistarn This is ready for review, I've added the example covering both the single file model case (sklearn) and a tarball for tensorflow. Thanks for you time!

File: docs/samples/uri/README.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+Apply the CRD,
+```bash
+kubectl apply -f sklearn_uri.yaml
+```
+Expected Output
+```
+$ inferenceservice.serving.kubeflow.org/sklearn-from-uri created
+```
+### Run a prediction
+Now, if everything went according to plan you should be able to hit the endpoint exposing the model we just uploaded.
```
Comment (yuzisun) [inline docs/samples/uri/README.md:60]: the ingress access instruction has been updated recently, can you help take a look at [tensorflow example](https://github.com/kubeflow/kfserving/tree/master/docs/samples/tensorflow#run-a-prediction) and change accordingly ?

File: python/kfserving/kfserving/storage.py [hunk 3]
```diff
@@ -206,6 +214,46 @@ def _download_local(uri, out_dir=None):
...
+            if response.status_code != 200:
+                raise RuntimeError("URI: %s returned a %s response code." % (uri, response.status_code))
+            if ext == '.zip' and not response.headers.get('Content-Type', '').startswith('application/zip'):
+                raise RuntimeError("URI: %s did not respond with \'Content-Type\': \'application/zip\'" % (uri))
+            elif not response.headers.get('Content-Type', '').startswith('application/octet-stream'):
+                raise RuntimeError("URI: %s did not respond with \'Content-Type\': \'application/octet-stream\'" % (uri))
+
+            # TODO: server-side gzip in Content-Encoding
+            # https://requests.readthedocs.io/en/master/user/quickstart/?highlight=.raw#raw-response-content
+            if ext in ['.tgz', '.gz']:
```
[the remark this answers is not in the document]
Comment (tduffy000) [reply python/kfserving/kfserving/storage.py:237]: thanks for digging this up! I was struggling with how to do this, the assumption that the file extension tells us what kind of file it actually is isn't very robust (and potentially a security hole), so I will update it.

File: docs/samples/uri/README.md (hunk 2 above)
Comment (tduffy000) [reply docs/samples/uri/README.md:60]: yes, will follow the posted example and update it

File: pkg/apis/serving/v1alpha2/utils.go [hunk 4]
```diff
@@ -82,16 +82,18 @@ func validateStorageURI(storageURI string) error {
 		return nil
 	}
 
-	// one of the prefixes we know?
-	for _, prefix := range SupportedStorageURIPrefixList {
-		if strings.HasPrefix(storageURI, prefix) {
+	// need to verify Azure Blob first, because it uses http(s):// prefix
```
[the remark this answers is not in the document]
Comment (tduffy000) [reply pkg/apis/serving/v1alpha2/utils.go:85]: yes, good point, will update!

File: docs/samples/uri/input.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+    "instances": [
+      [6.8,  2.8,  4.8,  1.4],
+      [6.0,  3.4,  4.5,  1.6]
+    ]
+  }
```
[the remark this answers is not in the document]
Comment (tduffy000) [reply docs/samples/uri/input.json]: https://github.com/kubeflow/kfserving/pull/979/commits/cfd872de0608a00ec3bfff364e500d66c51b44ad

File: docs/samples/uri/tensorflow.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+apiVersion: serving.kubeflow.org/v1alpha2
+kind: InferenceService
+metadata:
+  name: tensorflow-from-uri
+spec:
+  default:
+    predictor:
+      tensorflow:
+        storageUri: https://raw.githubusercontent.com/tduffy000/kfserving-uri-examples/master/tensorflow/frozen/model_artifacts.tar.gz
```
[the remark this answers is not in the document]
Comment (tduffy000) [reply docs/samples/uri/tensorflow.yaml]: https://github.com/kubeflow/kfserving/pull/979/commits/cfd872de0608a00ec3bfff364e500d66c51b44ad

File: pkg/apis/serving/v1alpha2/utils.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tduffy000) [reply pkg/apis/serving/v1alpha2/utils.go:85]: updated https://github.com/kubeflow/kfserving/pull/979/commits/bd542ffd5b67ade2bfde6a62d30b8b7433bf9d0e

File: python/kfserving/kfserving/storage.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tduffy000) [reply python/kfserving/kfserving/storage.py]: updated https://github.com/kubeflow/kfserving/pull/979/commits/56242ed25d2f0175ae5a57f2c48f5bf32701bc4f

**Comment by ellistarn** [comment]
I just want to chime in that this is great work! Thank you!

**Comment by yuzisun** [comment]
@tduffy000 Can you help do a rebase ?

**Comment by yuzisun** [comment]
Looks like we are keep getting this error and katib is making a fix here https://github.com/kubeflow/katib/pull/1316
```
ERROR: (gcloud.builds.submit) [Errno 104] Connection reset by peer
This may be due to network connectivity issues. Please check your network settings, and the status of the service you are trying to reach.
```

**Comment by tduffy000** [comment]
thanks @yuzisun , lmk if I can do anything to help out.

**Comment by yuzisun** [comment]
Keeps failing on gcloud auth login.. one more time.. We need the fix to add retries.

/retest

**Comment by jinchihe** [comment]
The teardown cluster step is passed now, but the e2e steps failed.
```
_______________ ERROR collecting transformer/test_transformer.py _______________
ImportError while importing test module '/root/go/src/github.com/kubeflow/kfserving/test/e2e/transformer/test_transformer.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.6/importlib/__init__.py:126: in import_module
   return _bootstrap._gcd_import(name[level:], package, level)
transformer/test_transformer.py:18: in <module>
   from kfserving import KFServingClient
E   ModuleNotFoundError: No module named 'kfserving'
```

**Comment by jinchihe** [comment]
That's strange, I reviewed code changes, should not cause the E2E test issue. Try again.

/retest

**Comment by jinchihe** [comment]
Failed again when build image, guess that's caused by gcloud build?
```
ERROR: (gcloud.builds.submit) [Errno 104] Connection reset by peer
This may be due to network connectivity issues. Please check your network settings, and the status of the service you are trying to reach
```

**Comment by yuzisun** [comment]
@tduffy000 Sorry for the delay of fixing the CI issues, can you help rebase again now?

**Comment by tduffy000** [comment]
No worries @yuzisun , thanks for all your help :smile:

**Comment by yuzisun** [comment]
omg! finally! @tduffy000 Thanks for your great contribution! Sorry for the long delayed mergy.

/lgtm
/approve

**Pull request merged**